File: backend/app/schemas/scenario.py
"""Scenario schemas for API validation"""
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, Field, field_validator


class WorldStateUpdate(BaseModel):
    """Schema for updating world state"""
    timeline: str | None = None
    location: str | None = None
    current_time: str | None = None
    active_events: list[str] | None = None
    world_facts: list[str] | None = None


class ScenarioBase(BaseModel):
    """Base scenario schema"""
    name: str = Field(..., min_length=1, max_length=200)
    description: str = Field(..., min_length=1)
    purpose: str = Field(default="", description="시나리오의 목적/목표")


class ScenarioCreate(ScenarioBase):
    """Schema for creating a new scenario"""
    world_state: dict = Field(default_factory=dict)
    is_public: bool = False


class ScenarioUpdate(BaseModel):
    """Schema for updating a scenario (all fields optional)"""
    name: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = Field(None, min_length=1)
    purpose: str | None = None
    world_state: dict | None = None
    is_public: bool | None = None


class ScenarioResponse(ScenarioBase):
    """Schema for scenario responses"""
    id: UUID
    user_id: UUID
    world_state: dict = Field(default_factory=dict)
    is_public: bool = False
    is_mine: bool = False
    character_avatars: list[dict] = Field(default_factory=list, description="캐릭터 이름+아바타 목록")
    created_at: datetime
    updated_at: datetime

    @field_validator("world_state", mode="before")
    @classmethod
    def ensure_world_state_dict(cls, v):
        if v is None:
            return {}
        return v

    model_config = {"from_attributes": True}


class ScenarioCharacterAdd(BaseModel):
    """Schema for adding a character to a scenario"""
    character_id: UUID


class ScenarioAutoGenerateRequest(BaseModel):
    """Schema for auto-generating scenario from source work"""
    source_work: str = Field(..., min_length=1, max_length=200, description="원작 작품명")


class CharacterInScenarioResponse(BaseModel):
    """Schema for character in scenario response"""
    id: UUID
    name: str
    avatar_url: str | None
    
    model_config = {"from_attributes": True}

File: frontend/lib/data/models/scenario.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'scenario.freezed.dart';
part 'scenario.g.dart';

@freezed
class Scenario with _$Scenario {
  const factory Scenario({
    required String id,
    required String userId,
    required String name,
    required String description,
    @Default('') String purpose,
    required Map<String, dynamic> worldState,
    @Default(false) bool isPublic,
    @Default(false) bool isMine,
    @Default([]) List<Map<String, dynamic>> characterAvatars,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = _Scenario;

  factory Scenario.fromJson(Map<String, dynamic> json) =>
      _$ScenarioFromJson(json);
}

@freezed
class ScenarioCreate with _$ScenarioCreate {
  const factory ScenarioCreate({
    required String name,
    required String description,
    @Default('') String purpose,
    @Default({}) Map<String, dynamic> worldState,
    @Default(false) bool isPublic,
  }) = _ScenarioCreate;

  factory ScenarioCreate.fromJson(Map<String, dynamic> json) =>
      _$ScenarioCreateFromJson(json);

  @override
  Map<String, dynamic> toJson() => {
        'name': name,
        'description': description,
        'purpose': purpose,
        'world_state': worldState,
        'is_public': isPublic,
      };
}

@freezed
class ScenarioUpdate with _$ScenarioUpdate {
  const factory ScenarioUpdate({
    String? name,
    String? description,
    String? purpose,
    Map<String, dynamic>? worldState,
    bool? isPublic,
  }) = _ScenarioUpdate;

  factory ScenarioUpdate.fromJson(Map<String, dynamic> json) =>
      _$ScenarioUpdateFromJson(json);

  @override
  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{};
    if (name != null) json['name'] = name;
    if (description != null) json['description'] = description;
    if (purpose != null) json['purpose'] = purpose;
    if (worldState != null) json['world_state'] = worldState;
    if (isPublic != null) json['is_public'] = isPublic;
    return json;
  }
}
